src/socket.rs

```diff
@@ -11,53 +11,72 @@ use critical_section::Mutex;
 use no_std_net::SocketAddr;
 use num_enum::{IntoPrimitive, TryFromPrimitive};
 
-const WAKER_INIT: Option<(Waker, i32)> = None;
-static SOCKET_WAKERS: Mutex<
-    RefCell<[Option<(Waker, i32)>; nrfxlib_sys::NRF_MODEM_MAX_SOCKET_COUNT as usize]>,
-> = Mutex::new(RefCell::new(
-    [WAKER_INIT; nrfxlib_sys::NRF_MODEM_MAX_SOCKET_COUNT as usize],
-));
+// use 16 slots for wakers instead of 8, which is the max number of sockets allowed, so they
+// are not overwritten when split into their rx/tx counterparts and run in separate tasks.
+const WAKER_SLOTS: usize = (nrfxlib_sys::NRF_MODEM_MAX_SOCKET_COUNT * 2) as usize;
+const WAKER_INIT: Option<(Waker, i32, i32)> = None;
+
+static SOCKET_WAKERS: Mutex<RefCell<[Option<(Waker, i32, i32)>; WAKER_SLOTS]>> =
+    Mutex::new(RefCell::new([WAKER_INIT; WAKER_SLOTS]));
 
 pub(crate) fn wake_sockets() {
     critical_section::with(|cs| {
         SOCKET_WAKERS
             .borrow_ref_mut(cs)
             .iter_mut()
             .for_each(|waker| {
-                if let Some((waker, _)) = waker.take() {
+                if let Some((waker, _, _)) = waker.take() {
                     waker.wake()
                 }
             })
     });
 }
 
-fn register_socket_waker(waker: Waker, socket_fd: i32) {
+fn register_socket_waker(waker: Waker, socket_fd: i32, socket_dir: i32) {
     critical_section::with(|cs| {
         // Get the wakers
         let mut wakers = SOCKET_WAKERS.borrow_ref_mut(cs);
 
         // Search for an empty spot or a spot that already stores the waker for the socket
-        let empty_waker = wakers
-            .iter_mut()
-            .find(|waker| waker.is_none() || waker.as_ref().map(|(_, fd)| *fd) == Some(socket_fd));
+        let empty_waker = wakers.iter_mut().find(|waker| {
+            waker.is_none()
+                || waker.as_ref().map(|(_, fd, dir)| (*fd, *dir)) == Some((socket_fd, socket_dir))
+        });
 
         if let Some(empty_waker) = empty_waker {
             // In principle we should always have an empty spot and run this code
-            *empty_waker = Some((waker, socket_fd));
+            *empty_waker = Some((waker, socket_fd, socket_dir));
         } else {
             // It shouldn't ever happen, but if there's no empty spot, we just evict the first socket
             // That socket will just reregister itself
             wakers
                 .first_mut()
                 .unwrap()
-                .replace((waker, socket_fd))
+                .replace((waker, socket_fd, socket_dir))
                 .unwrap()
                 .0
                 .wake();
         }
     });
 }
 
+#[derive(Clone, Copy, Debug)]
```

Comment:
If we want to compare the enum directly, then we probably want PartialEq and Eq here too